File: algorithms/InsertionSort.cpp
#include "Driver.h"

// Best-case: O(n)
// Worst-case (AND AVERAGE!): O(n^2)
// Note: internal loop has (n^2)/4 iterations
// Extra space: O(1)

template <typename T>
void insertionSort(std::vector<T>& input) 
{
  for (int i = 1; i < input.size(); ++i) {
    T key = input[i];
    int j = i - 1;
    while (j >= 0 && input[j] > key) {
      std::swap(input[j], input[j + 1]);
      --j;
    }
    input[j + 1] = key;
  }
}

int main(int argn, char* argv[]) {
  runMethod(argn, argv, insertionSort<int>);
  return 0;
}

File: algorithms/sorting/java/InsertionSort.java
class InsertionSort implements SortMethod
{
  @Override
  public void sort(int[] input)
  {
    for (int i = 1; i < input.length; ++i) {
      int key = input[i];
      int j = i - 1;
      while (j >= 0 && input[j] > key) {
        int temp = input[j];
        input[j] = input[j + 1];
        input[j + 1] = temp;
        --j;
      }
      input[j + 1] = key;
    }
  }

  public static void main(String[] argv)
  {
    Driver.run(new InsertionSort());
  }
}
